prusti-viper/src/encoder/encoder.rs

```diff
@@ -374,16 +373,13 @@ impl<'v, 'r, 'a, 'tcx> Encoder<'v, 'r, 'a, 'tcx> {
             if self_ty.is_self() {
                 if let Some(ty) = tymap[0].get(self_ty) {
                     if let Some(id) = tcx.trait_of_item(proc_def_id) {
-                        let proc_name = tcx.item_name(proc_def_id).to_string();
+                        let proc_name = tcx.item_name(proc_def_id).as_symbol();
                         tcx.for_each_relevant_impl(id, ty, |impl_id| {
-                            for assoc_item in tcx.associated_items(impl_id) {
-                                if assoc_item.name == proc_name {
-                                    if let Some(spec) = self.get_spec_by_def_id(assoc_item.def_id) {
-                                        impl_spec = spec.clone();
-                                    } else {
-                                        debug!("Procedure {:?} has no specification", assoc_item.def_id);
-                                    }
-                                    break;
+                            if let Some(item) = self.env().get_assoc_item(impl_id, proc_name) {
```

Comment:
I meant to move `for_each_relevant_impl` as well to a `Environment` method.

If I understand correctly, lines 377-378 iterate over all implementations of a trait and selects one associated item. So, this could become a `get_associated_item_impl(trait_id, ty, def_id) -> Option<...>` method in Environment. Using it, it would be explicit that `impl_spec = ...` cannot be executed more than once.